### ISSUE TITLE
OpenZwave add on isn't saving config to persistent storage

### ISSUE THREAD

**Issue by foxy82** [home-assistant/hassio-addons#1361]
The new openzwave add on isn't saving the OpenZwave data to persistent storage.

Inside the docker instead of writing the data to /data/ozw/config it is writing it to: /run/s6/services/zwave/config

**Comment by foxy82** [comment]
This could be related to this: https://github.com/OpenZWave/qt-openzwave/issues/75

A suggested work around would be to get the Docker to add the data into the persistent storage on startup.

**Comment by posixx** [comment]
Ah ok it is a hassio issue, i was relating it to qt-openzwave. It would be very helpful to have a solution for this!

**Comment by kpine** [comment]
Just to clarify, the root cause is the above mentioned qt-openzwave issue. It is a problem seen in this addon because it is overriding the default config path. Most Docker users would not see this problem because they are using the default config path.

**Comment by foxy82** [comment]
I disagree most docker users would mount a volume to save config. Otherwise each time you restart you will lose your zwave network which for battery devices could take 24 hours or more to come back. Even the qt-openzwave docs tell you to run their version in a Docker passing in a config volume: https://github.com/OpenZWave/qt-openzwave/blob/master/README.md

Even if people don't know it most (all) docker users of this add-on should want persistent storage.

**Comment by kpine** [comment]
I'm not sure what you are talking about. I said nothing like that.

**Comment by kpine** [comment]
Let me try to explain again. I never said docker users aren't using or don't want persistent storage. I said they don't have this problem because they _are_ using persistent storage.

1. The qt-openzwave version of the container sets the [WORKDIR ](https://github.com/OpenZWave/qt-openzwave/blob/47490b7bfc96535a34d1f101e0d98f0e338fe269/Docker/Dockerfile#L86)to `/opt/ozw`. The entrypoint script executes from the workdir. This means the current directory when the containers starts is `/opt/ozw`.
2. In https://github.com/OpenZWave/qt-openzwave/issues/75 I reported that the user-dir and config-dir command line arguments are ignored. Instead, the config and user directories are set to `./config`. `./config` is the same as `<current directory>/config`.

Put 1 and 2 together, and the user directory is always going to be `/opt/ozw/config` for the qt-openzwave version of the container (`/opt/ozw` + `./config`). The user directory is where the cache file is saved.

The example for the qt-openzwave container maps the persistent volume to `/opt/ozw/config` (well, there's an error in the run command, I think it will still persist anyways, just in a subdir, the docker compose example is correct). The example does not instruct you to override the config or user directories. So if you follow the instructions, which I presume most people are, the config/user directories are the same as what is selected by ozwdaemon, so you get your persistent cache file.

As for the addon... it has its own Dockerfile and run scripts that result in different behavior described above.

I am assuming the addon has a persistent volume at [`/data/ozw/config`](https://github.com/home-assistant/hassio-addons/blob/ecb8ce98b2cc5170d5025a0eb64f37a6b8b20cfc/zwave/rootfs/etc/services.d/zwave/run#L5). If this assumption is untrue, well then there is something to fix.

The run script is starting ozwdaemon with [`--config-dir /data/ozw/config --user-dir /data/ozw/config`](https://github.com/home-assistant/hassio-addons/blob/ecb8ce98b2cc5170d5025a0eb64f37a6b8b20cfc/zwave/rootfs/etc/services.d/zwave/run#L22), which tells ozwdaemon to save the user and config files there. Since it's a persistent location, the user/cache files are presume persistent.

Instead, because of https://github.com/OpenZWave/qt-openzwave/issues/75, the config/user directory settings are ignored and ozwdaemon ends up saving the cache to `<current directory/config`, as explained above. Based on your report, the current directory must be `/run/s6/services/zwave`, because that is where you have observed the config files going, and since that is not a persistent volume they will be lost.

If the addon does not want to wait for upstream to fix the problem, the workaround would be to modify the run script to change the working directory to `/data/ozw` prior to running ozwdaemon. ozwdaemon would pick `./config` as the location, which would result in a config/user directory setting of `/data/ozw/config`, and that is the persistent location.

**Comment by foxy82** [comment]
Thanks that workaround suggestion is useful.

Fyi - I've submitted a pr to fix the qt-openzwave documentation error you mention: https://github.com/OpenZWave/qt-openzwave/pull/87/files

**Comment by uSpike** [comment]
I think the fix for this is to update the versions in https://github.com/home-assistant/hassio-addons/blob/master/zwave/build.json to incorporate the fixes in https://github.com/OpenZWave/qt-openzwave/issues/75

**Comment by kpine** [comment]
Might want to wait at least until this issue is confirmed to be fixed: https://github.com/OpenZWave/qt-openzwave/issues/97

**Comment by frenck** [comment]
I've fixed this on the add-on side today, running locally for testing now (adding in some more fixes and features).
Expected to land soon.

**Comment by posixx** [comment]
Great news Frenck!
